train.py

```diff
@@ -1,21 +1,95 @@
+import time
+import torch
+
+
+def train_single_epoch(epoch, model, train_loader, optimizer, criterion, device):
+	"""
+	Train single epoch
+	"""
+	model.train()
+	epoch_start_time = time.time()
+	total_loss=0.
+	for i, batch in enumerate(iter(train_loader)):
+
+		img, target = batch
+		img = img.to(device)
+		target = target.to(device)
+		
+		optimizer.zero_grad()
+		#TODO: introduce for loop to make sentence
+		output, attentions, _ = model(img, target)
+
+		loss = criterion(output, target)
+		loss.backward()
+
+		torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=0.25)
+
+		optimizer.step()
+
+		total_loss += loss.item()
+
+		log_interval = 100
+
+		if i % log_interval == 0 and i > 0:
+			cur_loss = total_loss / log_interval
+			elapsed = time.time() - epoch_start_time
+			print(f'| epoch {epoch:3d} | {i:5d}/{len(train_loader):5d} batches | lr {lr:.4f} | ms/batch {elapsed * 1000 / log_interval:5.2f} | loss {cur_loss:5.2f} | ppl {math.exp(cur_loss):8.2f}')
+			start_time=time.time()
+
+def evaluate(model,dataiter):
+	model.eval()
+
+	total_loss = 0.
+
+	with torch.no_grad():
+		for idx, batch in enumerate(dataiter):
+			img, target = batch
+			img = img.to(device)
+			target = target.to(device)
+			#TODO: Adapt this piece of code to Encoder and decoder implementation
+			features = model.encoder...
+			output, attentions, _ = model.decoder.generate_caption(features, vocab=dataset.vocab)
+
+			caption = ' '.join(output)
```

Comment:
We can discuss this, not sure if the decoder needs to output all the caption at once but it’s not really important

Reply:
In any way, I think that the same output variable can't be used with a join for the caption retrieval and at the same time in the criterion function since criterion function is expecting a tensor with like word probabilities and the function join is expecting a list of strings? Don't you think?

Reply:
The output of the decoder can be a tensor, but the output of the generate_caption should be a list of words, thus the join.